File: src/scripts/extrair_dps_emissor.py
import os
import re
import json
import time
import requests
from bs4 import BeautifulSoup
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.serialization import pkcs12

PFX_PATH = os.environ.get("CERT_PFX_PATH", r"C:\Users\macar\OneDrive\Desktop\PISSARO VISTORIA AUTOMOTIVA LTDA 1010944009.pfx")
PASSWORD = os.environ.get("CERT_PASSWORD", "1234").encode()

OUTPUT_JSON = "notas-emissor-nacional.json"
OUTPUT_CSV = "notas-emissor-nacional.csv"
KEY_PATH = "nfse-conferencia/key.pem"
CERT_PATH = "nfse-conferencia/cert.pem"


def load_cert():
    with open(PFX_PATH, "rb") as f:
        pfx_data = f.read()
    private_key, certificate, _ = pkcs12.load_key_and_certificates(pfx_data, PASSWORD)

    os.makedirs(os.path.dirname(KEY_PATH), exist_ok=True)
    with open(KEY_PATH, "wb") as f:
        f.write(private_key.private_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PrivateFormat.TraditionalOpenSSL,
            encryption_algorithm=serialization.NoEncryption()
        ))

    with open(CERT_PATH, "wb") as f:
        f.write(certificate.public_bytes(serialization.Encoding.PEM))

    return (CERT_PATH, KEY_PATH)


def extract_dps(detail_html: str) -> str:
    soup = BeautifulSoup(detail_html, "html.parser")

    # Encontra a seção "Identificação do DPS" e pega o primeiro
    # <div class="form-group"> com label "Número"
    dps_section = soup.find(lambda tag: tag.name and tag.get_text(strip=True).lower() == "identificação do dps")
    if dps_section:
        for div in dps_section.find_all_next("div", class_="form-group"):
            label = div.find("label", class_="control-label")
            if label and label.get_text(strip=True).lower() == "número":
                value_span = div.find("span", class_="form-control-static")
                if value_span:
                    return value_span.get_text(strip=True)

    # fallback geral: procura o primeiro div com label "Número" + span com número
    for div in soup.find_all("div", class_="form-group"):
        label = div.find("label", class_="control-label")
        if label and label.get_text(strip=True).lower() == "número":
            value_span = div.find("span", class_="form-control-static")
            if value_span:
                return value_span.get_text(strip=True)
    return ""


def extract_nps_from_href(href: str) -> str:
    # href termina com chave de acesso; nNFSe fica no final da chave
    chave = href.split("/")[-1]
    match = re.search(r"0{6}(\d{3})\d{14}$", chave)
    return str(int(match.group(1))) if match else ""


def main():
    cert = load_cert()
    session = requests.Session()
    session.cert = cert

    print("Autenticando...")
    r = session.get("https://certificado.nfse.gov.br/EmissorNacional/Certificado", timeout=60)
    print("Login:", r.status_code, r.url)

    if "Dashboard" not in r.url and "Login" in r.url:
        raise Exception("Não conseguiu autenticar. Verifique o certificado.")

    notas = []
    page_num = 1
    while True:
        url = f"https://www.nfse.gov.br/EmissorNacional/Notas/Emitidas?pg={page_num}&datainicio=01/01/2026&datafim=10/09/2026&page=1"
        print(f"\nPágina {page_num}: {url}")
        r = session.get(url, timeout=60)
        if r.status_code != 200:
            print("  status:", r.status_code)
            break

        soup = BeautifulSoup(r.text, "html.parser")
        tabela = soup.find("table", {"class": "table"})
        if not tabela:
            print("  tabela não encontrada")
            break

        linhas = tabela.find("tbody")
        if not linhas:
            print("  tbody não encontrado")
            break

        trs = linhas.find_all("tr")
        print(f"  {len(trs)} linhas")
        if not trs:
            break

        for tr in trs:
            link = tr.find("a", href=re.compile(r"/EmissorNacional/Notas/Visualizar/Index/"))
            if not link:
                continue

            href = link.get("href", "")
            nfse_number = extract_nps_from_href(href)
            tds = tr.find_all("td")
            data = tds[0].get_text(strip=True) if tds else ""
            tomador = tds[1].get_text(strip=True) if len(tds) > 1 else ""

            cpf = re.search(r"\d{11}", tomador)
            cpf = cpf.group(0) if cpf else ""
            nome = re.sub(r"\s+", " ", re.sub(r"\d{11}", "", tomador)).strip()

            # acessa detalhe
            detail_url = f"https://www.nfse.gov.br{href}"
            print(f"  Nota {nfse_number}: {detail_url}")
            rd = session.get(detail_url, timeout=60)
            n_dps = extract_dps(rd.text) if rd.status_code == 200 else ""

            notas.append({
                "nfseNumber": nfse_number,
                "nDps": n_dps,
                "data": data,
                "cpf": cpf,
                "nome": nome,
                "chave": href.split("/")[-1],
            })

            print(f"    nNFSe {nfse_number} | nDPS {n_dps} | {data} | {cpf}")
            time.sleep(0.5)

        # Avança para a próxima página pelo parâmetro ?page=
        page_num += 1

    with open(OUTPUT_JSON, "w", encoding="utf-8") as f:
        json.dump(notas, f, indent=2, ensure_ascii=False)

    with open(OUTPUT_CSV, "w", encoding="utf-8") as f:
        f.write("nfseNumber,nDps,data,cpf,nome,chave\n")
        for n in notas:
            f.write(f"{n['nfseNumber']},{n['nDps']},{n['data']},{n['cpf']},\"{n['nome']}\",{n['chave']}\n")

    print(f"\nTotal notas extraídas: {len(notas)}")
    print(f"JSON: {OUTPUT_JSON}")
    print(f"CSV: {OUTPUT_CSV}")


if __name__ == "__main__":
    main()

File: src/scripts/extrair-dps-emissor.ts
import { chromium } from "playwright";
import * as fs from "fs";

interface NotaSite {
  nfseNumber: string;
  nDps: string;
  data: string;
  cpf: string;
  nome: string;
  chave: string;
}

function requiredEnv(name: string): string {
  const v = process.env[name];
  if (!v) throw new Error(`Env ${name} obrigatória`);
  return v;
}

function onlyDigits(s: string): string {
  return s.replace(/\D/g, "");
}

async function main() {
  const userDataDir = "emissor-user-data";
  fs.mkdirSync(userDataDir, { recursive: true });

  const context = await chromium.launchPersistentContext(userDataDir, {
    headless: true,
    channel: "chrome",
    viewport: { width: 1280, height: 900 },
    ignoreHTTPSErrors: true,
  });

  const page = await context.newPage();
  const notas: NotaSite[] = [];

  try {
    console.log("Abrindo login por certificado...");
    await page.goto("https://certificado.nfse.gov.br/EmissorNacional/Certificado", {
      waitUntil: "domcontentloaded",
      timeout: 120000,
    });
    await page.waitForTimeout(10000);
    await page.screenshot({ path: "emissor-cert.png" });

    console.log("Abrindo notas emitidas...");
    await page.goto("https://www.nfse.gov.br/EmissorNacional/Notas/Emitidas", {
      waitUntil: "domcontentloaded",
      timeout: 120000,
    });

    await page.waitForTimeout(5000);
    await page.screenshot({ path: "emissor-notas-emitidas.png" });
    fs.writeFileSync("emissor-notas-emitidas.html", await page.evaluate(() => document.body.innerHTML), "utf-8");

    // Enquanto houver próxima página
    let pagina = 1;
    for (;;) {
      console.log(`Página ${pagina}...`);

      const linhas = await page.locator("table tbody tr").all();
      console.log(`  ${linhas.length} linhas encontradas`);

      for (const linha of linhas) {
        const link = linha.locator("a[href*='/EmissorNacional/Notas/Visualizar/Index/']").first();
        const href = await link.getAttribute("href").catch(() => null);
        if (!href) continue;

        const nNFSeMatch = href.match(/(\d+)\d{14}$/);
        const nfseNumber = nNFSeMatch ? String(parseInt(nNFSeMatch[1], 10)) : "";

        const data = await linha.locator("td.td-data").first().textContent().catch(() => "") || "";
        const cpfRaw = await linha.locator("span.cpf").first().textContent().catch(() => "") || "";
        const nomeRaw = await linha.locator("td.td-texto-grande").first().textContent().catch(() => "") || "";

        // Abre detalhe em nova aba para extrair nDPS
        const detalhe = await context.newPage();
        try {
          await detalhe.goto(`https://www.nfse.gov.br${href}`, {
            waitUntil: "networkidle",
            timeout: 60000,
          });
          await detalhe.waitForTimeout(2000);

          const nDpsInput = detalhe.locator("#dps-numero, input[name*='DPS.Numero'], #NumeroDPS").first();
          const nDps = await nDpsInput.inputValue().catch(async () => {
            // tenta pelo label
            const label = await detalhe.locator("text=Identificação do DPS").first().isVisible().catch(() => false);
            if (label) {
              const input = detalhe.locator("input").filter({ has: detalhe.locator("xpath=preceding-sibling::*") }).first();
              return await input.inputValue().catch(() => "");
            }
            return "";
          });

          await detalhe.screenshot({ path: `emissor-detalhe-${nfseNumber}.png` });

          notas.push({
            nfseNumber,
            nDps: nDps.trim(),
            data: data.trim(),
            cpf: onlyDigits(cpfRaw),
            nome: nomeRaw.replace(/\s+/g, " ").trim(),
            chave: href.split("/").pop() || "",
          });

          console.log(`  Nota ${nfseNumber} | DPS ${nDps.trim()} | ${data.trim()} | ${onlyDigits(cpfRaw)}`);
        } catch (e: any) {
          console.error(`  Erro nota ${nfseNumber}: ${e.message}`);
        } finally {
          await detalhe.close();
        }
      }

      // Próxima página
      const proxima = page.locator("a[rel='next'], .pagination a:has-text('Próxima'), .pagination a:has-text('›')").first();
      const hasNext = await proxima.isVisible().catch(() => false);
      if (!hasNext) break;
      await proxima.click();
      await page.waitForTimeout(3000);
      pagina++;
    }

    fs.writeFileSync("notas-emissor-nacional.json", JSON.stringify(notas, null, 2), "utf-8");

    const header = "nfseNumber,nDps,data,cpf,nome,chave\n";
    const lines = notas.map((n) => `${n.nfseNumber},${n.nDps},${n.data},${n.cpf},"${n.nome}",${n.chave}`).join("\n");
    fs.writeFileSync("notas-emissor-nacional.csv", header + lines, "utf-8");

    console.log(`\nTotal notas extraídas: ${notas.length}`);
    console.log("JSON: notas-emissor-nacional.json");
    console.log("CSV: notas-emissor-nacional.csv");
  } catch (e: any) {
    console.error("Erro:", e.message);
    await page.screenshot({ path: "emissor-notas-erro.png" });
    fs.writeFileSync("emissor-notas-erro.html", await page.evaluate(() => document.body.innerHTML), "utf-8");
    process.exit(1);
  } finally {
    await context.close();
  }
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
